Enables cinder-driver to write host IPs to volume ACL to restrict access

rahman-lb opened the pull request:
This change enables cinder-driver to accept a list of the host IPs from connector properties,  and write this list to volume IP/ACL to restrict access.

-  Accept host IPs  form connector properties. 
-  Remove existing value (ALLOW_ANY)  from the given volume's IpAcl
-  Add IP addresses of the host to volume's IP-ACL upon connecting volume to a virtual machine.
- Remove IP addresses of the host from volume's IP-ACL upon disconnecting volume from a virtual machine.

rahman-lb committed:
This code change enables cinder-driver to accept a list of the host IPs from connector and write this list to volume IP/ACL to restrict access

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1033,17 +1036,18 @@ def _get_connection_properties(self, project_name, volume):
         return server_properties
 
-    def set_volume_acl(self, project_name, lightos_uuid, acl, etag):
+    def set_volume_acl(self, project_name, lightos_uuid, acl, ip_acl,etag):
```
nit: space after `ip_acl,`

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1062,23 +1066,39 @@ def __add_volume_acl(self, project_name, lightos_volname, acl_to_add):
+        if 'ALLOW_NONE' in ip_acl:
+            ip_acl.remove('ALLOW_NONE')
+        
+        ip_acl=set(ip_acl).union(set(host_ips))
```
I don't think we want to enable this behavior (IPACL) unconditionally. Better to add Cinder.conf configuration option `lightos_use_ipacl`, which can default to disabled (False), and once we are certain it's working, we'll make the default enabled (True).

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1033,17 +1036,18 @@ def _get_connection_properties(self, project_name, volume):
         )
 
-    def __add_volume_acl(self, project_name, lightos_volname, acl_to_add):
+    def __add_volume_acl(self, project_name, lightos_volname, acl_to_add,host_ips):
```
nit, space after `acl_to_add,`

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1062,23 +1066,39 @@ def __add_volume_acl(self, project_name, lightos_volname, acl_to_add):
             project_name,
             lightos_uuid,
-            acl,
+            acl,ip_acl,
```
nit: space after `acl,`

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1062,23 +1066,39 @@ def __add_volume_acl(self, project_name, lightos_volname, acl_to_add):
                 ''))
 
-    def add_volume_acl(self, project_name, volume, acl_to_add):
+    def add_volume_acl(self, project_name, volume, acl_to_add,host_ips):
```
same nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1137,10 +1156,34 @@ def __remove_volume_acl(
+                LOG.warning(
+                'Could not find matching ip %s in ip-acl %s of LightOS volume',
+                ip,
+                lightos_volname
```
the log message does not match the parameters (should be ip, ip_acl, lightos_volname) with the message format fixed appropriately)

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1137,10 +1156,34 @@ def __remove_volume_acl(
             project_name,
             lightos_uuid,
-            acl,
+            acl,ip_acl,
```
same nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1149,7 +1192,7 @@ def __overwrite_volume_acl(
             project_name,
             lightos_volname,
-            acl):
+            acl,host_ips):
```
same nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1168,20 +1211,20 @@ def __overwrite_volume_acl(
                 ''))
 
-    def remove_volume_acl(self, project_name, volume, acl_to_remove):
+    def remove_volume_acl(self, project_name, volume, acl_to_remove,host_ips):
```
same nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1168,20 +1211,20 @@ def __overwrite_volume_acl(
             project_name,
             lightos_uuid,
-            acl,
+            acl,host_ips,
```
nit: each arg on a different line

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1168,20 +1211,20 @@ def __overwrite_volume_acl(
             project_name,
             lightos_volname,
-            acl_to_remove)
+            acl_to_remove,host_ips)
```
same 2 nits: missing space and each arg on a line

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1191,17 +1234,17 @@ def remove_all_volume_acls(self, project_name, volume):
             project_name,
             lightos_volname,
-            ['ALLOW_NONE'])
+            ['ALLOW_NONE'],['ALLOW_NONE'])
```
nit: each arg on its own line

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1191,17 +1234,17 @@ def remove_all_volume_acls(self, project_name, volume):
+            ['ALLOW_NONE'],['ALLOW_NONE'])
 
-    def update_volume_acl(self, func, project_name, lightos_volname, acl):
+    def update_volume_acl(self, func, project_name, lightos_volname, acl,host_ips):
```
nit; space

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1191,17 +1234,17 @@ def remove_all_volume_acls(self, project_name, volume):
                 time.sleep(1)
             first_iteration = False
-            res = func(project_name, lightos_volname, acl)
+            res = func(project_name, lightos_volname, acl,host_ips)
```
nit: space

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1409,6 +1452,8 @@ def _delete_lightos_snapshot(self, project_name, snapshot_name):
         hostnqn = connector.get('nqn')
         found_dsc = connector.get('found_dsc')
+        host_ips = connector.get('host_ips',[])
+        LOG.info('Host current IP(s) are ',host_ips)
```
Which host? Better print the hostnqn at least, it will help debugging later

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1422,10 +1467,15 @@ def initialize_connection(self, volume, connector):
+        
+        if not host_ips:
+            msg = ('Connector (%s) did not find host IPs '
+                   'client, aborting' % (connector))
```
fix the error message, no need for client here

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1422,10 +1467,15 @@ def initialize_connection(self, volume, connector):
         lightos_volname = self._lightos_volname(volume)
         project_name = self._get_lightos_project_name(volume)
-        success = self.add_volume_acl(project_name, volume, hostnqn)
+        success = self.add_volume_acl(project_name, volume, hostnqn,host_ips)
```
nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1438,6 +1488,7 @@ def initialize_connection(self, volume, connector):
     def terminate_connection(self, volume, connector, **kwargs):
         force = 'force' in kwargs
         hostnqn = connector.get('nqn') if connector else None
+        host_ips = connector.get('host_ips',[]) 
```
note the `if connector else None` above -- you can be called with a None connector

rahman-lb replied:
If hostnqn is assigned value None, this code is not changing that behavior.  but host_ip will be empty list

muliby-lb replied:
Doesn't matter, you will crash on a None connector when calling `connector.get`. You can only call `connector.get` if you have a non-None connector.

rahman-lb replied:
got you, i added check for this

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1461,7 +1512,7 @@ def terminate_connection(self, volume, connector, **kwargs):
         lightos_volname = self._lightos_volname(volume)
         project_name = self._get_lightos_project_name(volume)
-        success = self.remove_volume_acl(project_name, volume, hostnqn)
+        success = self.remove_volume_acl(project_name, volume, hostnqn,host_ips)
```
nit: space

muliby-lb reviewed (commented):
Very good start. Please go over the review comments carefully and fix them.
I'm also missing a change to create volume to create each volume with a default `ALLOW_NONE` IPACL rather than the current permissive default

rahman-lb committed:
Adjusted code as per review comments,reading
 attribute from cinder.conf file for updating
 volume IP-ACL value if required.

Issue : LBM1-28409

adding needed space improve readability of the code

improve code readability by applying indentation

Added a function to user or not use ip/acl, also added check to avoid null pointer exception

use_ip_acl check when volume is removed from VM

fixed error message

create a volume with default IP-ACL value ALLOW_NONE or ALLOW_ANY based on lightos_use_ipacl in conf file

rahman-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1062,23 +1075,36 @@ def __add_volume_acl(self, project_name, lightos_volname, acl_to_add):
+           acl.append(acl_to_add)
 
+        ip_acl=[]    
+        if self.configuration.lightos_use_ipacl:
```
@muliby-lb  
Every volume is created with initial IP-ACL value "ALLOW_NONE",  and when volume is connected then this check will ensure if the flag **lightos_use_ipacl**  in cinder.conf is set to True or False, if this variable lightos_use_ipacl=true then code will write IPs to IP-ACL otherwise will set volume IP-ACL as ALLOW_ANY.

muliby-lb replied:
I'm not sure this is correct -- we are in *add*, not *overwrite*, so the IPACL should be added if use_ipacl is True and left untouched if use_ipacl is false.

rahman-lb replied:
@muliby-lb 
what would you recommend the volume's IP-ACL initial value when 
 -  lightos_use_ipacl=True   and  volume is created, but has not attached yet.
-   lightos_use_ipacl=False  and  volume is created, but has not attached yet.

Prior to this change ip-acl has always been created with initial value ALLOW_ANY. 

I added code similar to hostNQN,  a volume is created with initial IP/ACL value ALLOW_NONE and then later-on overwritten with either IP addresses or ALLOW_ANY based on the configuration of  lightos_use_ipacl=True|False

muliby-lb replied:
`lightos_use_ipacl = True` => initial IPACL is `ALLOW_NONE` or equivalent
`lightos_use_ipacl = False` => initial IPACL is `ALLOW_ANY`

rahman-lb commented:
@muliby-lb I updated the code and plz revisit the change.

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -77,7 +77,12 @@
+               ' an API endpoint response.'),
+    cfg.BoolOpt('lightos_use_ipacl',
+               default=True,
+               help='Use volume IP-ACL to restrict access by host,'
```
I suggest:
"IPACL work in conjunction with the standard NVME ACL. A host must be in both the IPACL and the ACL of a volume to access that volume. Cinder always sets the volume ACL. If lightos_use_ipacl is set to True, Cinder will also add the host's IP addresses to a volume's IPACL. If set to False, any IP address may access the volume. The default is True."

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1062,23 +1075,36 @@ def __add_volume_acl(self, project_name, lightos_volname, acl_to_add):
             project_name,
             lightos_uuid,
-            acl,
+            acl, ip_acl,
```
nit: each argument on its own line

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1089,13 +1115,13 @@ def add_volume_acl(self, project_name, volume, acl_to_add):
             project_name,
             lightos_volname,
-            acl_to_add)
+            acl_to_add,host_ips)
```
same nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1089,13 +1115,13 @@ def add_volume_acl(self, project_name, volume, acl_to_add):
             project_name,
             lightos_volname,
-            acl_to_remove):
+            acl_to_remove,host_ips):
```
same nit

muliby-lb commented on cinder/volume/drivers/lightos.py:
```diff
@@ -1137,10 +1162,34 @@ def __remove_volume_acl(
+        ip_acl = data.get('IPAcl')
+        if not ip_acl:
+            LOG.warning('Got LightOS volume without IP ACL?! data: %s', data)
+            return False
```
This code should only be executed if use_ipacl (suggest you put it in a function that returns immediately if use_ipacl is False)

muliby-lb reviewed (commented):
I added a few more comments. Please squash and fixup commits and resolve comments which were handled and I'll rereview.

yuval-lb commented:
looks good to me
